File: src/main/management/windows.ts
import path from 'path';
import { homedir, EOL } from 'os';
import {
  CLIMAT_HOME_DIR_NAME,
  moveManifestToClimatHome,
  removeAliasSymlinks,
  removeToolchain,
} from './utils';
import fs from 'fs-extra';
import { warn } from '../output/theme';

const join = path.win32.join;
export const CLIMAT_HOME = join(homedir(), CLIMAT_HOME_DIR_NAME);
export const TOOLCHAIN_HOME = join(CLIMAT_HOME, 'toolchains');
const climatBinPath = join(CLIMAT_HOME, 'bin');

function getBatchScript(name: string): string {
  return `@echo off${EOL}` + `climat runGlobal "${name}" %*`;
}
function getAliasScript(name: string): string {
  return `@echo off${EOL}` + `%~dp0${name} %*`;
}

function getBatchFilePath(name: string): string {
  return join(climatBinPath, `${name}.bat`);
}

export async function windowsInstall(
  manifest: string,
  name: string,
  aliases: string[],
): Promise<void> {
  await moveManifestToClimatHome(TOOLCHAIN_HOME, manifest, name);

  const batchFilePath = getBatchFilePath(name);
  await fs.ensureDir(climatBinPath);
  await fs.writeFile(batchFilePath, getBatchScript(name));

  const aliasScript = getAliasScript(name);
  await Promise.all(
    aliases.map((alias) => fs.writeFile(getBatchFilePath(alias), aliasScript)),
  );

  // TODO: add path automatically
  // This requires writing to the Windows registry
  // Was attempted before but it was deleted (see 608553d9073a08df9e2329c0e3d97354cca4b4e7)
  if (!process.env.PATH?.includes(climatBinPath)) {
    console.warn(warn(`Please add '${climatBinPath}' to system PATH`));
  }
}

export async function windowsUninstall(name: string): Promise<void> {
  await removeAliasSymlinks(name, climatBinPath, TOOLCHAIN_HOME, '.bat');
  await removeToolchain(TOOLCHAIN_HOME, name);
  await fs.rm(getBatchFilePath(name));
}

export async function windowsPurge(): Promise<void> {
  await fs.rm(CLIMAT_HOME, {
    recursive: true,
  });
}

File: src/main/kotlin/management/Windows.kt
package com.climat.management

import NodeJS.get
import com.climat.Fs
import com.climat.Path.win32
import com.climat.jsObjectOf
import com.climat.output.warn
import kotlinx.coroutines.await
import os.EOL
import os.homedir
import process
import kotlin.js.Promise

val WINDOWS_CLIMAT_HOME = win32.join(homedir(), CLIMAT_HOME_DIR_NAME)
val WINDOWS_TOOLCHAIN_HOME = win32.join(WINDOWS_CLIMAT_HOME, "toolchains")

private val climatBinPath = win32.join(WINDOWS_CLIMAT_HOME, "bin")

private fun getBatchScript(name: String): String = "@echo off${EOL}climat runGlobal \"$name\" %*"

private fun getAliasScript(name: String): String = "@echo off$EOL%~dp0$name %*"

private fun getBatchFilePath(name: String): String = win32.join(climatBinPath, "$name.bat")

suspend fun windowsInstall(
    manifest: String,
    name: String,
    aliases: Array<String>,
) {
    moveManifestToClimatHome(WINDOWS_TOOLCHAIN_HOME, manifest, name)

    val batchFilePath = getBatchFilePath(name)
    Fs.ensureDir(climatBinPath).await()

    Fs.writeFile(batchFilePath, getBatchScript(name)).await()

    val aliasScript = getAliasScript(name)

    Promise.all(
        aliases.map { Fs.writeFile(getBatchFilePath(it), aliasScript) }.toTypedArray(),
    ).await()

    // TODO: add path automatically
    // This requires writing to the Windows registry
    // Was attempted before but it was deleted (see 608553d9073a08df9e2329c0e3d97354cca4b4e7)

    if (process.env["PATH"]?.contains(climatBinPath) != true) {
        console.warn(warn("Please add '$climatBinPath' to system PATH"))
    }
}

suspend fun windowsUninstall(name: String) {
    removeAliasSymlinks(name, climatBinPath, WINDOWS_TOOLCHAIN_HOME, ".bat")
    removeToolchain(WINDOWS_TOOLCHAIN_HOME, name)
    Fs.rm(getBatchFilePath(name)).await()
}

suspend fun windowsPurge() {
    Fs.rm(
        WINDOWS_CLIMAT_HOME,
        jsObjectOf(
            "recursive" to true,
        ),
    ).await()
}
